### PR TITLE
Iso scattering

### PR DESCRIPTION
**Pull request by gbarbalinardo**
Implement isotopic scattering.

### CODE REVIEW

**Comment by gbarbalinardo** [comment]
@afiorent Is this branch good to be merged?

File: kaldo/controllers/isotopic.py [hunk 1]
```diff
@@ -56,11 +55,70 @@ def compute_isotopic_bw(phonons):
...
-        delta_omega=omegas-omegas[index_k,mu]
+        delta_omega=np.abs(omegas-omegas[index_k,mu])
         w2delta=omegas**2*curve(delta_omega,2*np.pi*sigma)
         bw=w2delta*g_per_mode/n_k_points
         isotopic_bw[index_k,mu]=(np.pi/2)*np.sum(bw[physical_mode])
 
+    return isotopic_bw
+
+@timeit
+def compute_isotopic_bw_condition(phonons,default_delta_threshold=3): # truncation of the gaussian delta after n sigma
```
Comment (gbarbalinardo) [inline kaldo/controllers/isotopic.py:66]: @afiorent this looks great! Do you think we could merge these two methods?

File: kaldo/controllers/isotopic.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+import numpy as np
+import ase.units as units
+from kaldo.helpers.tools import timeit
+from opt_einsum import contract
+from kaldo.helpers.logger import get_logger, log_size
+from kaldo.controllers.dirac_kernel import gaussian_delta, triangular_delta, lorentz_delta
+from ase.data.isotopes import download_isotope_data
+logging = get_logger()
+
+# @timeit
```
Comment (gbarbalinardo) [inline kaldo/controllers/isotopic.py:14]: Please remove the commented code.

File: kaldo/controllers/isotopic.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+#         delta_omega=np.abs(omegas-omegas[index_k,mu])
+#         w2delta=omegas**2*curve(delta_omega,2*np.pi*sigma)
+#         bw=w2delta*g_per_mode/n_k_points
+#         isotopic_bw[index_k,mu]=(np.pi/2)*np.sum(bw[physical_mode])
+#
+#     return isotopic_bw
+
+@timeit
+def compute_isotopic_bw(phonons,default_delta_threshold=3): # truncation of the gaussian delta after n sigma
+    speed_up = phonons.iso_speed_up
```
Comment (gbarbalinardo) [inline kaldo/controllers/isotopic.py:67]: Add method description.

File: kaldo/controllers/isotopic.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        g_per_mode=contract('ni,i->n',overlap,g_factor)
+        w2delta=omegas[condition]**2*curve(delta_omega[condition],2*np.pi*sigma)
+        bw=w2delta*g_per_mode/n_k_points
+        isotopic_bw[index_k,mu]=(np.pi/2)*np.sum(bw)
+
+    return isotopic_bw
+
+
+def calculate_base_sigma(velocity, cellinv, k_size):
+    #sigma: array (nk,nmodes)
```
Comment (gbarbalinardo) [inline kaldo/controllers/isotopic.py:129]: Please keep the format consistent with the rest of the code.

File: kaldo/controllers/isotopic.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            g_factor[list_of_atomic_numbers == element] = g_factor_dict[element]
+
+
+    return g_factor
+
+
+def legacy_database():
+    ## Legacy gfactor database. The isotopic database was downloaded with ase.data.isotopes on 20/03/2024.
+    # unstable elements have None as gfactor. Mostly elements with Z>92
+    database={1: 0.00011460742533846188, 2: 8.141017510851699e-08, 3: 0.001458820047621947, 4: 0.0, 5: 0.0013539153245004177, 6: 7.387230454537788e-05, 7: 1.8376657914224594e-05, 8: 3.3588025065260784e-05, 9: 0.0, 10: 0.0008278312160539562,\
```
Comment (gbarbalinardo) [inline kaldo/controllers/isotopic.py:182]: MInor, can you move this `database`  to a different file like a `isotopes.csv`.

File: kaldo/helpers/storage.py [hunk 6]
```diff
@@ -164,6 +164,9 @@ def get_folder_from_label(instance, label='', base_folder=None):
         if '<third_bandwidth>' in label:
             if instance.third_bandwidth is not None:
                 base_folder += '/tb_' + str(np.mean(instance.third_bandwidth))
+        if '<include_isotopes>' in label:
```
Comment (gbarbalinardo) [inline kaldo/helpers/storage.py:167]: nice!

File: kaldo/helpers/storage.py [hunk 7]
```diff
@@ -181,6 +184,9 @@ def get_folder_from_label(instance, label='', base_folder=None):
                     if '<finite_length_method>' in label:
                         if instance.finite_length_method is not None:
                             base_folder += '/fs' + str(instance.finite_length_method)
+        # if '<include_isotopes>' in label:
```
Comment (gbarbalinardo) [inline kaldo/helpers/storage.py:187]: remove if not needed.

File: kaldo/interfaces/shengbte_io.py [hunk 8]
```diff
@@ -201,11 +202,75 @@ def read_third_order_matrix_2(third_file, atoms, third_supercell, order='C'):
...
+    test_supercell = [int(x) for x in file.readline().split()] #the third supercell can be different to the second one
+    logging.info('Important: FORCECONSTANTS_3RD must be generated by d3q without recentering. Cell indeces must be from 0 to nrep, not [-nrep/2,nrep/2]')
+    if not (test_supercell[0]==supercell[0] and test_supercell[1]==supercell[1] and test_supercell[2]==supercell[2]):
+        print('third supercell not consistent with 3rd Forces file',test_supercell,supercell)
+
+    error_notation = 0
+    for i in range((3*n_unit_atoms)**3):
+        alpha,beta,gamma,atom_i,atom_j,atom_k=[int(x)-1 for x in file.readline().split()]
+        file.readline()
+        #print(alpha+1,beta+1,gamma+1,atom_i+1,atom_j+1,atom_k+1)
```
Comment (gbarbalinardo) [inline kaldo/interfaces/shengbte_io.py:239]: remove commented debug `print`, please

**Review by gbarbalinardo — COMMENTED** [review]
Terrific work @afiorent . Looks great to me. I added a few minor comments that should be addressed before merging.

File: kaldo/controllers/isotopic.py (hunk 5 above)
Comment (afiorent) [reply kaldo/controllers/isotopic.py]: Good idea, I just moved the database in a json file kaldo/controllers/legacy_dataset.json

**Review by niklundgren — APPROVED** [review]
Looks Great!

**Pull request merged**